core/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java

```diff
@@ -188,6 +204,50 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         return builder;
     }
 
+    public static MultiSearchResponse fromXContext(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    private static MultiSearchResponse.Item itemFromXContent(XContentParser parser) throws IOException {
+        // This parsing logic is complex here, because the multi search response is tricky.
+        // 1) The json objects inside the responses array are either a search response or a serialized exception
+        // 2) Each response json object gets a status field injected that ElasticsearchException.failureFromXContent(...) does not parse,
+        //    but SearchResponse.innerFromXContent(...) parses and then ignores. The status field is not needed to parse
+        //    the response item.
+
+        Token token = parser.nextToken();
+        assert token == Token.FIELD_NAME;
+        boolean statusBeenParsed = false;
+        String fieldName = parser.currentName();
+        if ("status".equals(fieldName)) {
+            token = parser.nextToken();
+            assert token == Token.VALUE_NUMBER;
+            token = parser.nextToken();
+            assert token == Token.FIELD_NAME;
+            fieldName = parser.currentName();
+            statusBeenParsed = true;
+        }
+
+        final Item item;
+        if ("error".equals(fieldName)) {
+            item = new Item(null, ElasticsearchException.failureFromXContent(parser));
+            if (statusBeenParsed) {
+                token = parser.nextToken();
+            } else {
+                token = parser.nextToken();
+                assert token == Token.FIELD_NAME;
+                token = parser.nextToken();
+                assert token == Token.VALUE_NUMBER;
+                token = parser.nextToken();
+            }
+            assert token == Token.END_OBJECT;
+        } else {
+            item = new Item(SearchResponse.innerFromXContent(parser), null);
+            assert parser.currentToken() == Token.END_OBJECT; // SearchResponse.innerFromXContent(...) consumes the entire json object
+        }
+        return item;
```

Comment:
looking at the code, it seems like this parsing method relies on keys ordering, but when I look at its test it shuffles the keys which means I am not reading it right. I guess the point is that `status` is ignored, and it will either ignored when it appears before any other element, or while parsing exception or search hits? Maybe that's what your comment above explained but I didn't get it.

Reply:
It does not really relies on key ordering but instead it checks the field one after the other, ignoring `status` if found as the first field or ignoring it again right after `error`, or let it be parsed by the `SearchResponse.innerFromXContent()`.

I agree that this parsing logic is not easy to do but I'm wondering if it could be more readable with the "usual" while loop that skip children on `status` field, and delegates to the appropriate parsing methods for `error` and search response?

Reply:
What @tlrx is saying is correct, it the parsing does not rely on ordering.

I think I can rewrite this code into a loop. However because`ElasticsearchException.failureFromXContent(...)` does not parse until end object and `SearchResponse#innerFromXContent(...)` does, the loop will have to immediately break in order to not read other search responses.

Reply:
sounds good to me Martijn thanks.